engines/osxnative/user.go

```diff
@@ -0,0 +1,149 @@
+package osxnative
+
+import (
+	"github.com/satori/go.uuid"
+	"os"
+	"path"
+	"strconv"
+)
+
+type user struct {
+	d    dscl
+	name string
+}
+
+func (u *user) create() error {
+	uid, err := u.getMaxUid()
+	if err != nil {
+		return err
+	}
+
+	staffGid, err := u.d.read("/Groups/staff", "PrimaryGroupID")
+	if err != nil {
+		u.delete()
+		return err
+	}
+
+	gid, err := strconv.Atoi(staffGid)
+	if err != nil {
+		return err
+	}
+
+	name := "worker-" + uuid.NewV4().String()
+	userPath := path.Join("/Users", name)
+
+	err = u.d.create(userPath)
+	if err != nil {
+		return err
+	}
+
+	// We set the uid, then check if it is unique, if
+	// not, increment and try again
+	duplicated := true
+	for duplicated {
+		uid++
+		strUid := strconv.FormatInt(int64(uid), 10)
+
+		err = u.d.create(userPath, "uid", strUid)
+		if err != nil {
+			u.delete()
+			return err
+		}
+
+		duplicated, err = u.isDuplicatedId("uid", uid)
+		if err != nil {
+			u.delete()
+			return err
+		}
+	}
+
+	err = u.d.create(userPath, "PrimaryGroupID", staffGid)
+	if err != nil {
+		u.delete()
+		return err
+	}
+
+	err = u.d.create(userPath, "NFSHomeDirectory", userPath)
+	if err != nil {
+		u.delete()
+		return err
+	}
+
+	err = os.MkdirAll(userPath, 0700)
+	if err != nil {
+		u.delete()
+		return err
+	}
+
+	err = os.Chown(userPath, uid, gid)
+	if err != nil {
+		u.delete()
+		return err
+	}
+
+	u.name = name
+
+	return nil
+}
+
+func (u *user) delete() error {
+	if u.name == "" {
+		return nil
+	}
+
+	userPath := path.Join("/Users", u.name)
+	err := u.d.delete(userPath)
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(userPath)
+	u.name = ""
+	return err
+}
+
+// return the next uid available
+func (u *user) getMaxUid() (int, error) {
+	uids, err := u.d.list("/Users", "uid")
+
+	if err != nil {
+		return -1, err
+	}
+
+	var maxUid int = 0
+	for _, entry := range uids {
+		uid, err := strconv.Atoi(entry[1])
```

Comment:
Oh, I'm totally cool with checking for errors, I was more so just curious for my own knowledge.